Added swap functionality to server client:

kyle-bradley opened the pull request:
Added a new chat command /swap based on the discussion in the associated issue: 
https://github.com/ddnet/ddnet/issues/1103

Swaps the character's position with another tee using the existing Save/Load functionality. 

Restrictions:
- Has to be in a team similar to the save/load feature

Possible additions:
- Time penalty

**Disclaimer**
This is my first contribution to the project. I tried my best to apply the correct coding standards and understand the purpose of each class/method. Please do provide feedback if used incorrectly. 

![Initiate Swap](https://user-images.githubusercontent.com/25198124/109426245-46ff0a80-79f5-11eb-935a-2b228a72695e.JPG)
![Pre-Acceptance](https://user-images.githubusercontent.com/25198124/109426249-4bc3be80-79f5-11eb-986a-eea9c6b84334.JPG)
![Post-Swap](https://user-images.githubusercontent.com/25198124/109426251-4e261880-79f5-11eb-80a3-772e23a23013.JPG)

## Checklist
- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

Commit:
Added swap functionality to server client:
- New chat command
- Swaps character's position with other tee using existing Save/Load methods.

edg-l commented on src/game/ddracechat.h:
```diff
@@ -27,6 +27,7 @@ CHAT_COMMAND("dnd", "", CFGFLAG_CHAT | CFGFLAG_SERVER | CFGFLAG_NONTEEHISTORIC,
 CHAT_COMMAND("mapinfo", "?r[map]", CFGFLAG_CHAT | CFGFLAG_SERVER, ConMapInfo, this, "Show info about the map with name r gives (current map by default)")
 CHAT_COMMAND("timeout", "?s[code]", CFGFLAG_CHAT | CFGFLAG_SERVER, ConTimeout, this, "Set timeout protection code s")
 CHAT_COMMAND("practice", "?i['0'|'1']", CFGFLAG_CHAT | CFGFLAG_SERVER, ConPractice, this, "Enable cheats (currently only /rescue) for your current team's run, but you can't earn a rank")
+CHAT_COMMAND("swap", "?r[player name]", CFGFLAG_CHAT | CFGFLAG_SERVER, ConSwap, this, "Swaps your position with a teammates")
```
teammate*

A better description maybe:
"Request to swap your tee with another team member"

houseme-kyle replied:
Much better - done :)

edg-l commented on src/game/server/ddracechat.cpp:
```diff
@@ -674,6 +674,100 @@ void CGameContext::ConPractice(IConsole::IResult *pResult, void *pUserData)
+		}
+	}
+
+	bool TargetNotFound = TargetClientId < 0;
```
I don't know the others' opinion, but we don't really use temporary booleans for ifs usually, specially if their meaning is obvious.

houseme-kyle replied:
Sure, it's a preference as I feel it makes it more readable. Happy to change it.

edg-l reviewed (commented):
Thanks for your first contribution 👍🏻 

I think a time penalty should be added, to prevent any cheats, I also think the command should have a cooldown, maybe 1 minute.

houseme-kyle commented:
> Thanks for your first contribution 👍🏻
> 
> I think a time penalty should be added, to prevent any cheats, I also think the command should have a cooldown, maybe 1 minute.

I'll send an update with the time penalty added now. RE the cooldown, I thought ProcessSpamProtection was a generic way of handling that?

fokkonaut commented:
Should also work in team 0 i think

fokkonaut commented on src/game/server/ddracechat.cpp:
```diff
@@ -674,6 +674,100 @@ void CGameContext::ConPractice(IConsole::IResult *pResult, void *pUserData)
+
+	CPlayer *pSwapPlayer = pSelf->m_apPlayers[TargetClientId];
+	if(!pSwapPlayer)
+	{ //Not sure what could cause this state.
```
nothing :)

houseme-kyle replied:
Removed

houseme-kyle commented:
> Should also work in team 0 i think

I'm worried there may be unforeseen consequences that people might be able to abuse. I thought it would be safer to release it in teams only first. Can always open it to team 0 in the future.

def- commented:
Thanks for the contribution! Great to see new people working on DDNet.

Mac build failure looks unrelated, so just ignore it.

In teams this replaces saving and loading again to get a player swapped, but it's not as flexible since you can't swap with someone from another team. I think providing it team-only is ok as a first try. Other opinions?

edg-l commented:
> 
> 
> Thanks for the contribution! Great to see new people working on DDNet.
> 
> Mac build failure looks unrelated, so just ignore it.
> 
> In teams this replaces saving and loading again to get a player swapped, but it's not as flexible since you can't swap with someone from another team. I think providing it team-only is ok as a first try. Other opinions?

I agree with sticking to team only for now.

Commit:
- Added timeout penalty
- Applied PR feedback

Commit:
Removed space

houseme-kyle commented:
I've made the penalty 5 minutes. Should be meaningful enough that can be used for speed runs while not ruining a person's run completely. 

Easy enough to change so opinions are welcome. 

Regarding the implementation of the penalty, I did write in the discord looking for ideas. I went with the above as it was the simplest.

Commit:
Fixed for formatting check

def- commented:
Instead of 5 min maybe use `sv_savegames_delay`

heinrich5991 commented:
I'd actually like to see a one minute waiting period like on a /save/load cycle. Other opinions?

def- commented:
Right, otherwise people could use this to quickly increase their team time. Or do you mean to actually stop the team from playing for 1 minute?

heinrich5991 commented:
I meant actually stopping the team from playing for one minute, to discourage this as a quick feature, instead encouraging people to try the part.

houseme-kyle commented:
> I meant actually stopping the team from playing for one minute, to discourage this as a quick feature, instead encouraging people to try the part.

I agree with penalizing the team for using /swap to prevent it from being used for speed-runs. What are the pros/cons of freezing them vs adding a time penalty of 1 minute / 5 minutes?

heinrich5991 commented:
> I feel the latter would be a better user experience.

I absolutely agree that it'd be a better user experience for the user using /swap. I think we should consider the second-level effects too, though: Enabling quick swap could lead to people routinely swapping because one person can do the part faster, instead of letting the weaker person try the part. I'd like to avoid that.

Learath2 commented:
Mh, artificially hindering the experience sounds very meh to me. What is bad with people swapping to do the parts they are faster at?

A freeze of this kind sounds like an implementation hurdle just begging to cause issues. We can do something like a deep freeze of all tees which might cause issues with edge cases like deep freeze sticking on for tees inside deep freeze or many other things I can't imagine. We can ignore all their inputs for a while but it could cause really annoying fails if people save while moving or hooked above. Given we want to see this in team 0 too the above solutions don't even work.

To sum up we need to make sure it's actually desirable to stop these people from playing because coming up with what this freeze should consist of is all but easy.

heinrich5991 commented:
> Mh, artificially hindering the experience sounds very meh to me.

That's precisely what we're doing rn with saves.

Learath2 commented:
> > Mh, artificially hindering the experience sounds very meh to me.
> 
> That's precisely what we're doing rn with saves.

If you want to keep status quo for status quo's sake, sure but if we can remove one more annoyance while at it I would think we would want to do it

houseme-kyle commented:
> > I feel the latter would be a better user experience.
> 
> I absolutely agree that it'd be a better user experience for the user using /swap. I think we should consider the second-level effects too, though: Enabling quick swap could lead to people routinely swapping because one person can do the part faster, instead of letting the weaker person try the part. I'd like to avoid that.

The intention of /swap is to allow people who have been stuck at a part to progress as a "last resort". It is actually meant to encourage learning e.g. a good player would be more willing to play/teach a bad player if they aren't potentially stuck for hours should the person come stuck on a section. My understanding is that there is a hidden feature already where a more advanced player would use /save /load to achieve this result. 

To address the concern there needs to be a penalty. Options can range from:
- No points at all (like practice)
- Small-time penalty like 1 minute
- Significant time penalty like 10 minutes

I am in favor of the latter so that this can never be used to save time. 

My concern with freezing people is that you limit the penalty you can give while the outcome is still the same.

Learath2 commented:
I disagree, if we make this more costly than the save swap, it'll just not be used. The idea was to make a feature out of the save swap.

I think @heinrich5991 doesn't want people to use it often so he wants it to be as annoying as the save swap. E.g. you can't play while waiting for the penalty. Which might be a good idea to stop them from using it too often but it has a lot of implementation pitfalls IMHO

houseme-kyle commented:
> I disagree, if we make this more costly than the save swap, it'll just not be used. The idea was to make a feature out of the save swap.
> 
> I think @heinrich5991 doesn't want people to use it often so he wants it to be as annoying as the save swap. E.g. you can't play while waiting for the penalty. Which might be a good idea to stop them from using it too often but it has a lot of implementation pitfalls IMHO

How about the middle-ground between both options then? 30-second freeze and a 5-minute time penalty?

The same consequence can apply for both /save and /swap to make it fair. Whether it is the above or another variation. 

At the end of the day, I feel that if certain people aren't happy with the time penalty for swap then it isn't for them. For me, it is meant to help people experience a map and learn it not for runs where you are worried about time.

fokkonaut commented on src/game/server/ddracechat.cpp:
```diff
@@ -674,6 +674,95 @@ void CGameContext::ConPractice(IConsole::IResult *pResult, void *pUserData)
+		pSelf->Console()->Print(IConsole::OUTPUT_LEVEL_STANDARD, "swap", "Swap request sent. The player needs to confirm it.");
+
+		char aBuf[512];
+		str_format(aBuf, sizeof(aBuf), "'%s' has requested to swap positions with you. Type /swap to confirm the swap.", pSelf->Server()->ClientName(pResult->m_ClientID));
```
"Type '/swap %s' to confirm the swap."

houseme-kyle replied:
Good catch will update

def- commented:
save/load still adds 1 minute, if not that's a bug.

Commit:
- Added name to message
- Fixed format issue

Commit:
Fixed missing variable

Chairn commented:
> 
> 
> I dont think there should be any penalty. You should just add a timer to use this command again.
> I remember save/load used to add time too, but it got removed so I think this should work the same.

I agree, 1 minute time penalty + a delay on the two swapping tees to prevent them swapping again like 1 minute as well, but they can still play during this delay

houseme-kyle commented:
Okay, so a few differing opinions but it seems we have landed with a time penalty and a cooldown period for the feature. I personally think a higher penalty than 1 minute to make sure it isn't used in speed-runs.

Final thoughts on:
- 5-minute time penalty
- 1-minute cooldown

heinrich5991 commented:
I continue to think that this should block the tees or the whole team from movement for a minute (if it's the whole team, the team could also do some more swaps in the time span). Since we absolutely never remove any features, I think it's more prudent to give /swap the same limitations that /save/load has. We can relax those later. We can probably not make them stricter later.

@Learath2 I think it's not impossible to implement this correctly, even first-try-ish.

Making the penalty higher than /save/load isn't really worthwhile either, I think; if the same effect can be achieved with a /load/save cycle, the more user-unfriendly method would be preferred if the team cared about time.

Learath2 commented:
> I continue to think that this should block the tees or the whole team from movement for a minute (if it's the whole team, the team could also do some more swaps in the time span). Since we absolutely never remove any features, I think it's more prudent to give /swap the same limitations that /save/load has. We can relax those later. We can probably not make them stricter later.

I see your point with us not being able to go back on this, but still is this not just trying to make it annoying so that people don't use it? I thought the point of promoting this to a first class feature was to show people we are fine with it and reduce annoyance.

> @Learath2 I think it's not impossible to implement this correctly, even first-try-ish.

Well I'm all ears. The only way that I'm completely sure would work is simulating a save/load and that doesn't "scale" into a future where we might want to let people do this in team0 too.

I see a valid concern with people just swapping all the time, but it's not like you can force someone to swap with you. So I can only see it being a big issue in team0 with pro players getting tilted and saying swap or I leave.

kamillentee commented:
So I want to add some arguments:

1.  should of course be useable in team 0, otherwise it's useless (teams already got a swap feature)
2. I agree with heinrich5991, by default tees should be prevented from moving  for a certain time (or add a time penality which sounds easier. It gives people the ability to swap in the air tho:/ ) but add a way to let mappers reduce the time to zero. It won't touch old maps too much while giving mappers the possibility to build parts around that feature.
3. having a cooldown of 5 minutes (maybe even 10) looks good, because otherwise people will pressure harder to swap if one cannot do the part fast enough. This alone already adresses most of the concerns with people swapping all the time for speedrunning.


But can we please talk about the time penalty (and for  save/load the 1 minute waiting time too). If I understand it correctly reason for those were the imagination of preventing people from saving time in speedruns because one can do the part faster (or people benefiting from the strong swap too much). 
And here is my problem with it: 1 full minute is centuries for ddrace speedrunners. They often will kill and restart already when there is 3 seconds lost. For example look at this speedrun video and see how much progress they make in a random 1 minute timespan https://www.youtube.com/watch?v=w-xHI7_ZYGc The time penalty should  barely disencourage people from using it to be faster than to have a huge time waiting for no reason. 
Speedrunners by the way usually race with partners with a simliar skill level, I doubt they can save too much time with swapping then. If we imagine the worst case of unbalanced skill level (steam player versus 10 year experience tw crack), swapping can probably significantly reduce the time needed here, but the pro player is more likely to be faster with his dummy anyway than with a random beginner.
So my suggestion ist to reduce the time penalty to only 10 seconds ( and the waiting time too for save/load), should be enough to cover all possibilities to be faster with swap or save. If not name me  examples  how to misuse that please

houseme-kyle commented:
Okay, I'm going to try to summarize this discussion and see where there is alignment and where there isn't.

**Core Design Principles**
1. /swap and /save /load to have the same functionality and limitations as each other
2. /swap shouldn't a feature that is frequently used and seen as a way to "speed-run" or discourage someone from trying
3. /swap should be encouraged as a way for players to become unblocked examples include:
* A player needs to leave and is in a critical position
* A player has attempted a part and rather than give up the other player can give it a go

I have a feeling most of us agree on the above points. The difference seems to be around how best to apply point number 2.

**Penalty Options**
There are three types of discouragement mentioned that can be applied:
1. Cooldown - feature can only be used every X minutes. 
* Seems to be general agreement on this. 
* Can I go with 5 minutes?
2. Freeze - the team is frozen for X minutes
* Not only a penalty of time but a nuisance so people will unlikely default to it unless absolutely needed
* There is a limit to the time penalty possible as you don't want to freeze people for too long
* **this doesn't appear to currently happen for /save /load**
* I'm in favour of keeping it this way or increasing it 30 seconds. I agree with Learath that making it more make cause unnecessary  annoyance
* Can I go with 0 minutes?
3. Time added - the team gets X minutes added to their score
* There does appear to be a penalty added currently. Either 30 seconds or 1 minute (TBC)
* I'm in favour of making this more than 1 minute to cement the idea that this isn't there to help with your run
* Can I go with 5 minutes?

So we can start narrowing down the scope, please tell me which point you disagree with and a suggested alternative. One big change from previous is that our understanding of how /save /load currently works appears to be wrong. 
![image](https://user-images.githubusercontent.com/25198124/110820804-1b530e80-8298-11eb-9b17-fb03b58381b8.png)

houseme-kyle commented:
@kamillentee 

1. The idea is to not have the first version for team 0. Only advanced and committed players know about using /save /load to achieve this and actually do so.
2. The intention is for the feature not to impact the gameplay at all but rather to swap the person behind the tee. How would a part be built around it?
3.  Great, seems like you also agree with point 1 then :)

kamillentee commented:
> 
> 1. /swap and /save /load to have the same functionality and limitations as each other
> 
No, swap should be useable in team 0 too. It's still a valid argument that having it in teams only does not bring any new oropportunities. I see that you want to limit it to teams *at first* and maybe expand it in the future. Why not now, nobody could name any possible cheats compared to team swap. One guy mentioned people could swap and finish for others: That's not wrong but already very easy possible and all are doing it. People who dont want to gift their rank, simply dont change their name right now and in future they will not swap either. If you really want to limit people from doing so you have to make rules against that and need to punish them.

> Cooldown?
> * Can I go with 5 minutes?
> 
Seems good

> Freeze people?
> * Can I go with 0 minutes?
> 
Not against that either

> Time penality?
> * I'm in favour of making this more than 1 minute to cement the idea that this isn't there to help with your run
> * Can I go with 5 minutes?
> 
People can misuse that to gain extra high times. And again how can you justify penalties over 1 minute, if 1 minute is already way to long to gain any time by swapping, Remember save/load is already possible and gives less of a penalty. My arguments from my post above still stand valid, so name me examples how could anybody save even 20 seconds with swapping, I just don't have any idea. In Combination with cooldown its even less usable for speedrunning. A solution should fit to the problem, this here is cracking a nut with a sledgehammer.

Learath2 commented:
Okay here is what I would be okay with.

1 minute time penalty: This is to keep it functionally the same as the save load swap.

1 minute nuisence: This I'm also okay with but I want to see implementation ideas for these because we can't just freeze people in t0. You need to make sure people can't interact with these people.

5 minute cooldown: Sounds like a good deterrant to remove the noob pressuring I mentioned earlier. However, this cripples it with respect to the power of save/load. Are we okay with that?

kamillentee commented:
> Cooldown 5 minutes

After further thinking about it: no we are not okay with that. In  maps that are typical for big groups it sometimes happens that a big group (lets say 20 people) alle fail except for this one guy they are dragging through the map (who failed every part before but it is how it is). 
Right now all hope for this guy and write something like: "You can do it", "We love you", "I believe in you", "You suck", "Hero <3" And if this one guy really manages to save, all will be extra relieved. With swap they will instantly put pressure on this guy to swap with someone better. So an idea might be to simulate the timeout switch (100seconds waiting time before swapping). It adds a little inconvenience, enough so people rather let the bad guy try it. And with that get rid of the time penalty since when you decide to switch you normally already wasted time trying the part many times. From registering the will to swap till it can be done finally is kind of a time penalty too.

heinrich5991 commented:
@kamillentee Thanks for putting into words what I fear from this change.

> You need to make sure people can't interact with these people.

Freezing your input in T0 might actually be the way to go; this is approximately what happens when you do the timeout solution.

edg-l commented:
I really think this feature should only be available to teams, otherwise players might just swap with a player who already finished in team 0, giving a finish for free with 0 effort.

In a team this is different because you can't just swap with someone at the end of the map, since you need to be with your teammates to reach the end.

Having this in team 0 really gives the incentive to be lazy and swap with someone at the end.

heinrich5991 commented:
@edg-l Another option would be to remove swapping in T0. (But right now, it's possible with timeouts.)

kamillentee commented:
@edg-l  

>One guy mentioned people could swap and finish for others: That's not wrong but already very easy possible and all are doing it. People who dont want to gift their rank, simply dont change their name right now and in future they will not swap either. If you really want to limit people from doing so you have to make rules against that and need to punish them.

Enforced rules against finishing for others works really well for KoG btw

@heinrich5991 Do you really think you want to limit a feature that is known and used almost 7 years.


So if this swap feature gets released as it is right it's completly useless. Team 0 who are waiting for a real swap feature since /save was implemented don't get anything at all. And teams will not benefit too much from it since they can save and load already *without* a cooldown but the same time penalty as the swap feature

corneum commented:
i didnt read the whole thing here but this is how i imagine it:
once you typed /swap a personalised vote appears on your teammate he presses f3 and you get to spawn where a countdown will show in the broadcast. 20-30 seconds because 1 minute is annoying. maybe even less because i dont see how this can be used for cheats except your teammate is good and does all parts or something.

for bigger teams you type /swap "nickname" and everybody gets to decide if the two of them get to swap or not.

idk if votes and broadcasts can be personalised but it would be cool anyway for /practice votes

houseme-kyle commented:
@kamillentee Great point and I think that is an excellent idea. A cooldown but in reverse. Might in fact solve most of the concerns we are trying to address.

Going back to the three penalty types
1) Cooldown changes to:
- The other player cannot accept the swap until X minutes have passed
- /Swap resets after 1 minute. Prevents calling /swap from the start to avoid cooldown
// I am in favour of 5 minutes. Long enough that the person might succeed and disincentive enough that people don't rely on it by default

2) Freeze penalty
- Think this can be avoided now given the above

3) Time penalty
- Think this can be avoided as /swap can't be used in speed-runs due to the waiting period.

houseme-kyle commented:
On extending this to /t0. Can we delay that discussion/decision until after this release as the next iteration? 

Having it for teams only is how /save /load currently behaves and releasing it on teams-only first doesn't preclude us from extending both to /t0 in the near future.

ardadem commented:
Something like sv_savegames_delay-10 sec penalty and 10 sec cooldown would be better imo. It will prevent quick swap and reduce annoyance.

heinrich5991 commented:
Yea, I think this new scheme is a good idea.

To recap:
There is a delay until the swap can be accepted (maybe 1 min). You have to accept it within a certain timeframe or it expires (maybe another 30 seconds?). Then you'll have a cooldown of say 5 min before you can suggest a swap again.

You'll also get a time penalty for swapping, maybe also 1 min like sv_savegames_delay.

houseme-kyle commented:
I feel like there is enough consensus on this item to go ahead with an update. 

As an FYI, the existing sv_savegames_delay doesn't affect penalty btw. It is a hard-coded value.
`int Time = m_Time + 60 * SERVER_TICK_SPEED;`

Going to use sv_savegames_delay for the cooldown value and create a new sv_savegames_penalty for this value. 

Seeing that you can play during this period, I think:
sv_savegames_delay = 3 minutes
sv_savegames_penalty = 1 minute (current)

Commit:
Updated to include penalty, cooldown and timeout

houseme-kyle commented:
![image](https://user-images.githubusercontent.com/25198124/111388544-2226ba80-86b8-11eb-862b-90959be4debf.png)

I'm going to do some more testing tomorrow - especially seeing that I did some refactoring on the team class. I thought I would get this out so long as the bigger hold-up previously has been on the implementation specifics.

Commit:
Style fixes

Commit:
Removed comments and added variable reset

Commit:
Merge pull request #5 from ddnet/master

Merge

Commit:
Edge case for player state when changing team

houseme-kyle commented:
Okay, I am pretty confident in the PR now. Did some more checks/tests today. Ready to be checked in.

def- committed:
Fix clang-format

By using clang-format-10 from AUR clang-format-static package

Commit:
Further tests and edge cases dealt with

Commit:
Merge branch 'feature-swap' of https://github.com/houseme-kyle/ddnet into feature-swap

def- commented:
If there are no complaints by this evening, I will submit this.

heinrich5991 reviewed (commented):
Looks mostly good.

I found some parts of the code hard to follow, it might be nice to add a comment what the `m_ClientSwapID` means. The bug that duplicates the save game penalty for old saves should be fixed.

houseme-kyle commented on src/game/server/teams.cpp:
```diff
@@ -675,6 +679,82 @@ void CGameTeams::OnFinish(CPlayer *Player, float Time, const char *pTimestamp)
 	}
 }
 
+void CGameTeams::RequestTeamSwap(CPlayer *Player, CPlayer *TargetPlayer, int Team)
```
Happy to make the change, but the other methods in the class are capitalized. Are they in an old format? Is there a reason this should be different?

e.g.
`CGameTeams::OnFinish(CPlayer *Player, float Time, const char *pTimestamp)`
`CGameTeams::GetCpCurrent(CPlayer *Player)`

Just want to improve my understanding of the naming conventions.

Learath2 replied:
Notice it's not `player` but `pPlayer`. We use a sort of hungarian notation.

`g_` "global"
`s_` static
`m_` member
`_p` pointer
`_a` array
`_l` this is rarer but "list"

houseme-kyle replied:
@Learath2 Sure, but the other methods have it as **CPlayer *Player.** To confirm, they are still in an old standard?

Learath2 replied:
They are not in standard, just coded carelessly or not reviewed strictly :) We'll go thru the entire repo sometime if #2945 ever gets settled.

houseme-kyle commented on src/game/server/teams.cpp:
```diff
@@ -675,6 +679,82 @@ void CGameTeams::OnFinish(CPlayer *Player, float Time, const char *pTimestamp)
+	m_LastSwap[Team] = Server()->Tick();
+}
+
+void CGameTeams::SwapTeamCharacters(CPlayer *Player, CPlayer *TargetPlayer, int Team)
```
Likewise

houseme-kyle commented on src/game/server/ddracechat.cpp:
```diff
@@ -674,6 +674,81 @@ void CGameContext::ConPractice(IConsole::IResult *pResult, void *pUserData)
+	int TargetClientId = -1;
+	for(int i = 0; i < MAX_CLIENTS; i++)
+	{
+		if(!str_comp(pName, pSelf->Server()->ClientName(i)))
```
I took this logic from the ConInvite function, so have to be honest I didn't test it extensively. If it can't find/match a player's name then it returns "Player Not Found". 

Not sure if you are referring to a different use-case.

heinrich5991 replied:
I'm wondering `Server()->ClientName(i)` returns if `i` is not a valid client ID. I think it might be something like `"(invalid)"` or so, making it impossible to swap with a player called `"(invalid)"`.

If you copied it from another function, that function is probably also broken in that way. Sorry that the codebase is broken in that way. :(

houseme-kyle replied:
Seems to be the opposite, it returns (invalid) or (connecting) if ClientID is in the wrong state or ClientID is invalid. 

My understanding is that the impact for a player with the named (invalid) would be an occasional failure to swap, as it matches on the joining ClientID rather than the player with the (invalid) name. 

Can't think of a straightforward solution given the way the function currently works. I think the use-case is unusual enough to ignore and look at a more permanent refactor down the line. One could be to return an empty string. Without knowing the larger implication I wouldn't want to include it in this PR.

heinrich5991 replied:
I agree that I wouldn't want to refactor the client name function. I think it'd still be worth working around it, by checking whether a player with that ID is online before trying to get their name. You could e.g. check if `pSelf->m_apPlayers[i]` is non-NULL for this before trying to compare the name.
```suggestion
		if(pSelf->m_apPlayers[i] && !str_comp(pName, pSelf->Server()->ClientName(i)))
```

houseme-kyle replied:
Thanks for the suggestion - my solution would not have been so elegant.  

I've tested it out locally and it works, so will commit it as is.

houseme-kyle commented on src/game/server/save.cpp:
```diff
@@ -145,7 +145,7 @@ void CSaveTee::Load(CCharacter *pChr, int Team)
 
 	if(m_Time)
-		pChr->m_StartTime = pChr->Server()->Tick() - m_Time;
+		pChr->m_StartTime = pChr->Server()->Tick() - m_Time - g_Config.m_SvSaveGamesPenalty * pChr->Server()->TickSpeed();
```
Good point, I wanted to change it to be on the load because I didn't like the idea of persisting the configured penalty. Also, the save function seems to be called each tick so ever so slightly less demanding. 

Anyway, moved it back to be on the save.

Commit:
PR feedback:
- Changed names of variables
- Changed config descriptions
- Fix for perma timeout

Commit:
Reverted test config

Commit:
Fixed names

Learath2 reviewed (approved):
Looks good to me aswell. Only concern I have remaining is whether we tested this in game thoroughly enough given it changes some team code.

houseme-kyle commented:
> Looks good to me aswell. Only concern I have remaining is whether we tested this in game thoroughly enough given it changes some team code.

Agreed that there is that risk. I really tried to test as many scenarios as possible. I don't know the different edge-cases so hard to be completely thorough.

Learath2 commented:
If you say you tested it as well as you could I'm fine with it. Just wanted to know you tested the team aspect too. We can deploy it on a single server to test all together when it's merged. +1 from me (as it is right now at 35991ea)

Commit:
Scenario of (invalid) or (connection) names breaking /swap

Co-authored-by: heinrich5991 <heinrich5991@gmail.com>

def- commented:
bors r+

Should I run it on GER2 first before rebuilding for other servers?

Learath2 commented:
> Should I run it on GER2 first before rebuilding for other servers?

Yes
